File: 2299-merge-nodes-in-between-zeros/merge-nodes-in-between-zeros.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def mergeNodes(self, head: Optional[ListNode]) -> Optional[ListNode]:
        dummy = ListNode(0)  
        result = dummy
        su = 0
        current = head.next  

        while current:
            if current.val == 0:
                result.next = ListNode(su)
                result = result.next
                su = 0
            else:
                su += current.val
            current = current.next

        return dummy.next
        

File: 2299-merge-nodes-in-between-zeros/merge-nodes-in-between-zeros.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    ListNode* mergeNodes(ListNode* head) {
        ListNode* curr = head->next; // Skip the first zero node
        ListNode* result = new ListNode(0); // Dummy node for results
        ListNode* temp = result; // Pointer to add nodes to result
        int sum = 0;

        while (curr != NULL) {
            if (curr->val != 0) {
                sum += curr->val;
            } else {
                temp->next = new ListNode(sum);
                temp = temp->next;
                sum = 0; // Reset the sum for the next segment
            }
            curr = curr->next;
        }

        return result->next; // Skip the dummy node
    }
};
